File: backend/app/api/v1/communities.py
from fastapi import APIRouter, Depends

from app.api.deps import get_community_service, get_current_user
from app.core.response import success
from app.middleware.jwt_auth import CurrentUser
from app.schemas.community import (
    CommunityCreateRequest,
    CommunityUpdateRequest,
    InviteCreateRequest,
    UpdateMemberRoleRequest,
)
from app.services.community_service import CommunityService

router = APIRouter(prefix="/communities", tags=["communities"])


@router.post("")
def create_community(
    body: CommunityCreateRequest,
    current_user: CurrentUser = Depends(get_current_user),
    service: CommunityService = Depends(get_community_service),
):
    data = service.create_community(
        current_user.user_id,
        name=body.name,
        description=body.description,
    )
    return success(data.model_dump())


@router.get("")
def list_communities(
    current_user: CurrentUser = Depends(get_current_user),
    service: CommunityService = Depends(get_community_service),
):
    data = service.list_communities(current_user.user_id)
    return success([item.model_dump() for item in data])


@router.get("/{code}")
def get_community(
    code: str,
    current_user: CurrentUser = Depends(get_current_user),
    service: CommunityService = Depends(get_community_service),
):
    data = service.get_community(current_user.user_id, code)
    return success(data.model_dump())


@router.patch("/{code}")
def update_community(
    code: str,
    body: CommunityUpdateRequest,
    current_user: CurrentUser = Depends(get_current_user),
    service: CommunityService = Depends(get_community_service),
):
    data = service.update_community(
        current_user.user_id,
        code,
        name=body.name,
        description=body.description,
    )
    return success(data.model_dump())


@router.delete("/{code}")
def delete_community(
    code: str,
    current_user: CurrentUser = Depends(get_current_user),
    service: CommunityService = Depends(get_community_service),
):
    service.delete_community(current_user.user_id, code)
    return success(message="社区已解散")


@router.get("/{code}/members")
def list_members(
    code: str,
    current_user: CurrentUser = Depends(get_current_user),
    service: CommunityService = Depends(get_community_service),
):
    data = service.get_community(current_user.user_id, code)
    return success([item.model_dump() for item in data.members])


@router.delete("/{code}/members/{user_code}")
def remove_member(
    code: str,
    user_code: str,
    current_user: CurrentUser = Depends(get_current_user),
    service: CommunityService = Depends(get_community_service),
):
    service.remove_member(current_user.user_id, code, user_code)
    return success(message="成员已移除")


@router.patch("/{code}/members/{user_code}")
def update_member_role(
    code: str,
    user_code: str,
    body: UpdateMemberRoleRequest,
    current_user: CurrentUser = Depends(get_current_user),
    service: CommunityService = Depends(get_community_service),
):
    service.update_member_role(current_user.user_id, code, user_code, body.role)
    return success(message="成员角色已更新")


@router.post("/{code}/invitations")
def create_invitation(
    code: str,
    body: InviteCreateRequest,
    current_user: CurrentUser = Depends(get_current_user),
    service: CommunityService = Depends(get_community_service),
):
    data = service.create_invitation(
        current_user.user_id,
        code,
        email=body.email,
    )
    return success(data.model_dump())


@router.get("/{code}/invitations")
def list_invitations(
    code: str,
    current_user: CurrentUser = Depends(get_current_user),
    service: CommunityService = Depends(get_community_service),
):
    data = service.list_invitations(current_user.user_id, code)
    return success([item.model_dump() for item in data])


@router.delete("/{code}/invitations/{invite_code}")
def revoke_invitation(
    code: str,
    invite_code: str,
    current_user: CurrentUser = Depends(get_current_user),
    service: CommunityService = Depends(get_community_service),
):
    service.revoke_invitation(current_user.user_id, code, invite_code)
    return success(message="邀请已撤销")


@router.get("/{code}/shared-pages")
def list_community_shared_pages(
    code: str,
    current_user: CurrentUser = Depends(get_current_user),
    service: CommunityService = Depends(get_community_service),
):
    data = service.list_community_shared_pages(current_user.user_id, code)
    return success([item.model_dump() for item in data])

File: frontend/src/lib/api/communities.ts
import type { SharedPageDto } from './pages';
import { apiRequest } from './client';

export type CommunitySummaryDto = {
  code: string;
  name: string;
  description: string | null;
  is_public: boolean;
  owner_code: string;
  owner_name: string;
  member_count: number;
  my_role: string | null;
  my_application_status: string | null;
  created_at: string;
};

export type CommunityMemberDto = {
  user_code: string;
  name: string;
  email: string;
  role: string;
  joined_at: string;
};

export type CommunityDetailDto = CommunitySummaryDto & {
  members: CommunityMemberDto[];
};

export type InvitationDto = {
  code: string;
  invitee_email: string;
  status: string;
  expires_at: string;
  invite_link: string;
  community_code?: string;
  community_name?: string;
  inviter_name?: string;
};

export type JoinApplicationDto = {
  code: string;
  status: string;
  message: string | null;
  applicant_code: string;
  applicant_name: string;
  applicant_email: string;
  community_code?: string;
  community_name?: string;
  created_at: string;
  reviewed_at: string | null;
};

export async function fetchCommunities(options?: { signal?: AbortSignal }) {
  return apiRequest<CommunitySummaryDto[]>('/communities', options);
}

export async function fetchPublicCommunities(options?: { signal?: AbortSignal }) {
  return apiRequest<CommunitySummaryDto[]>('/communities/public', options);
}

export async function fetchCommunity(code: string) {
  return apiRequest<CommunityDetailDto>(`/communities/${code}`);
}

export async function fetchCommunityMembers(code: string) {
  return apiRequest<CommunityMemberDto[]>(`/communities/${code}/members`);
}

export async function createCommunity(input: {
  name: string;
  description?: string | null;
  is_public?: boolean;
}) {
  return apiRequest<CommunitySummaryDto>('/communities', {
    method: 'POST',
    body: input,
  });
}

export async function updateCommunity(
  code: string,
  input: {
    name?: string;
    description?: string | null;
    is_public?: boolean;
  }
) {
  return apiRequest<CommunitySummaryDto>(`/communities/${code}`, {
    method: 'PATCH',
    body: input,
  });
}

export async function deleteCommunity(code: string) {
  await apiRequest<null>(`/communities/${code}`, { method: 'DELETE' });
}

export async function removeMember(communityCode: string, userCode: string) {
  await apiRequest<null>(`/communities/${communityCode}/members/${userCode}`, {
    method: 'DELETE',
  });
}

export async function updateMemberRole(
  communityCode: string,
  userCode: string,
  role: 'admin' | 'member'
) {
  await apiRequest<null>(`/communities/${communityCode}/members/${userCode}`, {
    method: 'PATCH',
    body: { role },
  });
}

export async function createInvitation(communityCode: string, email: string) {
  return apiRequest<InvitationDto>(`/communities/${communityCode}/invitations`, {
    method: 'POST',
    body: { email },
  });
}

export async function fetchInvitations(communityCode: string) {
  return apiRequest<InvitationDto[]>(`/communities/${communityCode}/invitations`);
}

export async function revokeInvitation(communityCode: string, inviteCode: string) {
  await apiRequest<null>(`/communities/${communityCode}/invitations/${inviteCode}`, {
    method: 'DELETE',
  });
}

export async function applyToJoinCommunity(
  communityCode: string,
  message?: string | null
) {
  return apiRequest<JoinApplicationDto>(`/communities/${communityCode}/applications`, {
    method: 'POST',
    body: { message: message ?? null },
  });
}

export async function fetchJoinApplications(communityCode: string) {
  return apiRequest<JoinApplicationDto[]>(`/communities/${communityCode}/applications`);
}

export async function approveJoinApplication(
  communityCode: string,
  applicationCode: string
) {
  return apiRequest<JoinApplicationDto>(
    `/communities/${communityCode}/applications/${applicationCode}/approve`,
    { method: 'POST' }
  );
}

export async function rejectJoinApplication(
  communityCode: string,
  applicationCode: string
) {
  return apiRequest<JoinApplicationDto>(
    `/communities/${communityCode}/applications/${applicationCode}/reject`,
    { method: 'POST' }
  );
}

export async function fetchMyInvitations() {
  return apiRequest<InvitationDto[]>('/invitations/mine');
}

export async function fetchInvitationPreview(inviteCode: string, token: string) {
  return apiRequest<InvitationDto>(
    `/invitations/${inviteCode}?token=${encodeURIComponent(token)}`,
    { auth: false }
  );
}

export async function acceptInvitation(inviteCode: string, token?: string) {
  return apiRequest<CommunitySummaryDto>('/invitations/accept', {
    method: 'POST',
    body: { invite_code: inviteCode, token: token ?? null },
  });
}

export async function fetchCommunitySharedPages(communityCode: string) {
  return apiRequest<SharedPageDto[]>(`/communities/${communityCode}/shared-pages`);
}
